Do not reflect default values

LeaVerou opened the pull request:
Implementation on the base class and sample usage on the `appearance` property of `<wa-button>`.

- Fixes #506
- Fixes https://github.com/shoelace-style/shoelace/discussions/2339
- Paves the way for a ton of improvements

This implements a new `default` descriptor for `@property` and
- wraps the [default converter](https://lit.dev/docs/components/properties/#conversion-type) with a check: if the value is equal to the default value (regardless of how it was set), it does not create an attribute (and removes it if it was already there). Since it imports the default converter and passes the value back to it, the implementation is very minimal, and quite robust.
- wraps the default getter to return the default value if the getter returns `null` or `undefined`.

The only DX "cost" for us is specifying the default value twice — which is actually only needed for the CE manifest, the actual functionality doesn't need it, so perhaps down the line we can just patch the CEM.

KonnorRogers commented:
Looks pretty good to me. There's one odd edge case where if a user sets the property to the same as the default value, it will not reflect. Don't know if its worth supporting this, but it is worth noting.


https://github.com/user-attachments/assets/7a942f2a-ef7a-4683-86b8-b7e8b448bf72

LeaVerou commented:
> Looks pretty good to me. There's one odd edge case where if a user sets the property to the same as the default value, it will not reflect. Don't know if its worth supporting this, but it is worth noting.

That is by design. It's a compromise we need to make, otherwise complexity increases very significantly.

I think it's actually a good thing though, it teaches people about default values and it follows the design principle of [preferring the shortest form when serializing](https://github.com/w3ctag/design-principles/issues/284#issuecomment-773611014).

My main gripe with the current implementation is that we need to specify the default values twice, just for the sake of CEM. But oh well.

LeaVerou commented:
One issue with this approach is specific to these examples: that if we take no attribute to mean "inherit", then there is no way to specify that a button within a callout should have a neutral color. So perhaps the default value for these should be `auto` or something like that.

claviska commented:
> That is by design. It's a compromise we need to make, otherwise complexity increases very significantly.

To confirm, have all existing component styles been updated to _not_ use reflection?

The main reason reflection was even added was to facilitate styling. At this point, it might be better to update the policy on when we want to reflect and update any remaining components that still do.

So, if styling is handled with classes and states now, what _is_ our reflection policy? 🤔

LeaVerou commented:
Btw I didn't merge this because I don't think the right approach is making `neutral` the default value, it's having an `auto` default value so you can still explicitly specify a `neutral` that does not reflect. And possibly updating the getter to return resolved values (and not `auto`). See #506 for the design discussion on that.

> > That is by design. It's a compromise we need to make, otherwise complexity increases very significantly.
> 
> To confirm, have all existing component styles been updated to _not_ use reflection?

This does not disable reflection — it only removes the attribute when it's equal to the default. And yes, that's something we can (and in most cases have, but it will need more testing) encode in CSS.

claviska commented:
> This does not disable reflection — it only removes the attribute when it's equal to the default. And yes, that's something we can (and in most cases have, but it will need more testing) encode in CSS.

I'm aware. My point being if defaults no longer reflect, it would bork a lot of styles if they haven't already been updated.

Thanks for clarifying. I'll keep an eye out for them in the CSS.

claviska commented:
My other point being: we're probably reflecting things without reason now, so we may want to revisit what we _are_ still reflecting since it's likely arbitrary if not for styling.

LeaVerou commented:
> > This does not disable reflection — it only removes the attribute when it's equal to the default. And yes, that's something we can (and in most cases have, but it will need more testing) encode in CSS.
> 
> I'm aware. My point being if defaults no longer reflect, it would bork a lot of styles if they haven't already been updated.

Gotcha. I think we might want to be deliberate about this and opt properties in (or at least provide a way to opt them out) rather than applying it to everything at once.

LeaVerou committed:
Do not reflect default values

- Add `default` descriptor
- Do not reflect attributes when equal to their default value
- Patch getter return value to return default value when empty
- Use it in button `appearance`
